packages/web/src/apollo/index.tsx

```diff
@@ -22,7 +22,16 @@ const ApolloProviderWithFetchConfig: React.FunctionComponent<{
   config?: Omit<ApolloClientOptions<InMemoryCache>, 'cache'>
 }> = ({ config = {}, children }) => {
   const { uri, headers } = useFetchConfig()
-  const { getToken, type: authProviderType, isAuthenticated } = useAuth()
+  const {
+    getToken,
+    type: authProviderType,
+    isAuthenticated,
+    loading,
+  } = useAuth()
+
+  if (loading) {
+    return null
```

Comment:
I don't want to stop rendering here. I want rendering to continue down to the router so it can show the `whileLoading()` stuff.

Reply:
https://github.com/redwoodjs/redwood/blob/42ec5eb1cbc8302fc9802e0b1d3b9068b91e7361/packages/router/src/router.tsx#L65-L67

Reply:
@Tobbe I tried putting in the same `whileLoading` logic on the public routes in router.tsx as well and even used a private route in this issue and in both cases I end up stuck on the "/" route even though the address bar shows "/after-auth" as I demonstrated in #1892 

So far this was the only thing that fixed the issue.

https://github.com/redwoodjs/redwood/blob/53349fd121f9198a8fdb5c0217cec8c4f8b6badd/packages/web/src/components/RedwoodProvider.tsx#L40-L45

Based on this comment from `0.21` someone has always wanted it removed but I've pretty much narrowed it down to the one thing causing both issues in #1892.

Reply:
Thanks for looking in to this @o0charlie0o What you're describing explains a few odd things I've noticed with auth state as well. So I'd love to get to the bottom of this.

15 minutes later...

So I've been digging through the code, and I think I might have a solution. I'm too tired right now to keep everything in my head, so I'll have to actually try it out to know if it'll work or not. I'll try to verify my idea tomorrow. In the meantime, can you help me by describing the absolute simplest way to reproduce this issue in a newly created RW app? Or if you can set up a repo I can clone that shows the problem. Thanks!

Reply:
Sure I'll try to do that tonight. I had an idea as well to use navigate inside the Auth0 callback rather than replaceState but haven't tried it yet and not very confident it'll work.

Reply:
@Tobbe Sorry this took so long. Got a little busy. This is the repo I've been using while testing this.

https://github.com/o0charlie0o/redwoodblog

Reply:
No worries, I've been pretty busy on my end as well! What kind of auth do I need to set up to play with that?

Reply:
@Tobbe I've been using Auth0. The variables are in the .env.example file.

Reply:
I have cloned your repo now, and reproduced the issue.
Next step for me is to see if I can fix it :)

Reply:
I've spent some time working on this. I have also talked to @peterp about it. I have a way forward, but it's a lot of work, and restructuring of the RW code. Got stuck on a timing issue inside the auth code. Peter said he would think about it, see if he could come up with a good solution.

Basically I've created a new package, @redwoodjs/history that lifts that stuff higher up, so it can be used from both auth and router.

Unfortunately I don't think I will have time to finish this in the current sprint.

Reply:
@Tobbe thanks for the update!

Reply:
Hi gents, confirming we'll be pausing PR merges for this Sprint by the end-of-day tomorrow. Tomorrow will be focused on the final review and QA for PRs in this Sprint. 

I'll move this as a priority for the next Sprint. Also, I'm working on improving our patch release process, which means hopefully we could get this out as a patch as soon as it's ready. TBD

Thanks for keeping this moving forward. Definitely important!

Reply:
Posted a WIP PR https://github.com/redwoodjs/redwood/pull/2212